### PR TITLE
subsystem benchmarks: add cpu profiling

### PR DESCRIPTION
**Pull request by AndreiEres**
Ready-to-merge version of https://github.com/paritytech/polkadot-sdk/pull/2601

- Added optional CPU profiling
- Updated instructions how to set up Prometheus, Pyroscope and Graphana
- Added a flamegraph dashboard
image

### CODE REVIEW

File: polkadot/node/subsystem-bench/docker/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+services:
+  grafana:
+    image: grafana/grafana-enterprise:latest
```
Comment (tdimitrov) [inline polkadot/node/subsystem-bench/docker/docker-compose.yml:3]: nit: Shouldn't we use an explicit version here?

File: polkadot/node/subsystem-bench/docker/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    container_name: grafana
+    restart: always
+    networks:
+      - subsystem-bench
+    ports:
+      - "3000:3000"
+
+  prometheus:
+    image: prom/prometheus:latest
```
Comment (tdimitrov) [inline polkadot/node/subsystem-bench/docker/docker-compose.yml:12]: And here

File: polkadot/node/subsystem-bench/docker/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      - ./prometheus:/etc/prometheus
+    extra_hosts:
+      - "host.docker.internal:host-gateway"
+    ports:
+      - "9090:9090"
+      - "9999:9999"
+
+  pyroscope:
+    container_name: pyroscope
+    image: grafana/pyroscope:latest
```
Comment (tdimitrov) [inline polkadot/node/subsystem-bench/docker/docker-compose.yml:27]: And here too

File: polkadot/node/subsystem-bench/docker/docker-compose.yml (hunk 1 above)
Comment (AndreiEres) [reply polkadot/node/subsystem-bench/docker/docker-compose.yml:3]: Why? I think that yml is going to be updated much rarer than the grafana (and others) version. And it's only for local development, if someone has particular preferences they can change it locally.

File: polkadot/node/subsystem-bench/src/subsystem-bench.rs [hunk 4]
```diff
@@ -76,12 +78,34 @@ struct BenchCli {
...
+	/// Enable CPU Profiling with Pyroscope
+	pub profile: bool,
+
+	#[clap(long, default_value_t = String::from("http://localhost:4040"))]
+	/// Pyroscope Server URL
+	pub pyroscope_url: String,
+
+	#[clap(long, default_value_t = 113)]
+	/// Pyroscope Sample Rate
+	pub pyroscope_sample_rate: u32,
```
Comment (alindima) [inline polkadot/node/subsystem-bench/src/subsystem-bench.rs:91]: we should have these two args require that `profile` is set. There should be a `requires` clap attribute that does that

File: polkadot/node/subsystem-bench/README.md [hunk 5]
```diff
@@ -192,8 +220,7 @@ view the test progress in real time by accessing [this link](http://localhost:30
 
 Now run
 `target/testnet/subsystem-bench test-sequence --path polkadot/node/subsystem-bench/examples/availability_read.yaml`
-and view the metrics in real time and spot differences between different  `n_valiator` values.
-  
+and view the metrics in real time and spot differences between different `n_valiator` values.
```
Comment (ordian) [inline polkadot/node/subsystem-bench/README.md:223]:
```suggestion
and view the metrics in real time and spot differences between different `n_validator` values.
```

File: polkadot/node/subsystem-bench/src/subsystem-bench.rs (hunk 4 above)
Comment (AndreiEres) [reply polkadot/node/subsystem-bench/src/subsystem-bench.rs:91]: Good catch!

**Review by alindima — APPROVED** [review]

File: polkadot/node/subsystem-bench/src/subsystem-bench.rs [hunk 6]
```diff
@@ -76,12 +78,34 @@ struct BenchCli {
...
 
+	#[clap(long, default_value_t = false)]
+	/// Enable CPU Profiling with Pyroscope
+	pub profile: bool,
+
+	#[clap(long, requires = "profile", default_value_t = String::from("http://localhost:4040"))]
+	/// Pyroscope Server URL
+	pub pyroscope_url: String,
+
+	#[clap(long, requires = "profile", default_value_t = 113)]
```
[the remark this answers is not in the document]
Comment (AndreiEres) [reply polkadot/node/subsystem-bench/src/subsystem-bench.rs:89]: It's the same value we use in the main polkadot's pyroscope. See: https://github.com/paritytech/polkadot-sdk/blob/ebe2aad6f0ae576a0e176f38a084fe7579f936dd/polkadot/cli/src/command.rs#L300

According to the PR it was added with, it has no a reason behind. Just a prime number, more than 100.

File: polkadot/node/subsystem-bench/docker/docker-compose.yml (hunk 1 above)
Comment (tdimitrov) [reply polkadot/node/subsystem-bench/docker/docker-compose.yml:3]: With `latest` you have no idea what version was used for a deployment, which is bad from reproducibility point of view. Also if a breaking change is introduced in any of the images we might run into issues.

But if you say it's just for local development and we don't care for the above - fine by me :)

**Review by tdimitrov — APPROVED** [review]

**Review by alexggh — APPROVED** [review]

**Pull request merged**